File: memory_example.c
#include <stdio.h>
#include <stdlib.h>

int main() {

    char *str = malloc(6); //allocates 6 bytes of memory to the variable str
    str[0] = 'h';
    str[1] = 'e';
    str[2] = 'l';
    str[3] = 'l';
    str[4] = 'o';
    str[5] = '\0'; //indicates end of a string
    //str now has a value, "hello"
    free(str); //frees up the memory used by str
    //str no longer has a value 
    //if we attempt to use it beyond this point the compiler will throw an error

    return 0;
}

File: memory_example.py
class Person:
    def __init__(self, age, gender) -> None:
        self.age = age
        self.gender = gender

john = Person(25, "male")
print(john)

# What you see:
str = "Hello World"

# # What your computer sees:
str = "Hello World" 
str_address = 0x1f94753b8b0 #Don't let this crazy string of numbers and letters spook you!
# #The computer recognizes the line above as an address in the physical memory of your computer
# #now that address has a value "Hello World", and can be used elsewhere

# print(str) #let's say we use our variable x for the last time here
# #at this point, the garbage collector sweeps through and cleans up the memory we allocated to str
# #str no longer exists in the physical memory of your computer